subsys/bluetooth/host/settings.c

```diff
@@ -215,6 +215,12 @@ static int commit(void)
 			BT_ERR("Unable to setup an identity address");
 			return err;
 		}
+
+		/* Make sure that the identity automatically generated by
+		 * bt_setup_id_addr is saved to persistent storage.
+		 * This only applies if settings_load was called without any
+		 * active identities in RAM. */
+		bt_settings_save_id();
```

Comment:
I recall @jhedberg saying there was some specific order to do this properly, or perhaps I missing something.

Reply:
Yes, that’s the point. This is being done only when resuming the bt_enable() sequence through the app calling settings_load(). The issue was that the default ID created by the stack was not being saved, which makes it very hard for apps to use privacy.

Reply:
The order of calling bt_id_create is documented to have a specific order.
The problem is however when bt_id_create is not called at all.
This default behavior is not working as I would expect it to be.

Consider these scenarios.

1# 1 id created, not stored, id will be restored (except it is not stored)
```
bt_enable
bt_settings_load
```
2#  - 2 ids created, not stored, ids will not be restored
```
bt_id_create
bt_id_create
bt_enable
bt_settings_load
```
3#  2 ids created, not stored, ids will be restored (except they are not stored)
```
bt_enable
bt_id_create
bt_id_create
bt_settings_load
```
4# 3 ids created, stored, ids will be restored
```
bt_enable
bt_settings_load
bt_id_create
bt_id_create
```
 ---

The current diff will change:
1# 1 id created, stored, id will be restored
```
bt_enable
bt_settings_load
```

I'm also wondering if 3# should be changed. The documentation for bt_id_create says:
```
 *  However,
 *  the new identity will only be stored persistently in flash when this API
 *  is used after bt_enable().
```
So either we need to change this behaviour as well, or update the documentation.

Reply:
Yep, I can't think of any reasons not to store in case of #1, that would break the stored bonding if we change the id on the next reboot.

Reply:
This would fix 3# as well:

```
                        BT_ERR("Unable to setup an identity address");
                        return err;
                }
+       }
 
-               /* Make sure that the identity automatically generated by
-                * bt_setup_id_addr is saved to persistent storage.
-                * This only applies if settings_load was called without any
-                * active identities in RAM. */
+       /* Make sure that the identities created by bt_id_create after
+        * bt_enable is saved to persistent storage. */
+       if (!atomic_test_bit(bt_dev.flags, BT_DEV_PRESET_ID)) {
                bt_settings_save_id();
        }

```

Reply:
+1 to this